doc/pcre2posix.3
.TH PCRE2POSIX 3 "27 November 2024" "PCRE2 10.49"
.SH NAME
PCRE2 - Perl-compatible regular expressions (revised API)
.SH "SYNOPSIS"
.rs
.sp
.B #include <pcre2posix.h>
.PP
.nf
.B int pcre2_regcomp(regex_t *\fIpreg\fP, const char *\fIpattern\fP,
.B "     int \fIcflags\fP);"
.sp
.B int pcre2_regexec(const regex_t *\fIpreg\fP, const char *\fIstring\fP,
.B "     size_t \fInmatch\fP, regmatch_t \fIpmatch\fP[], int \fIeflags\fP);"
.sp
.B "size_t pcre2_regerror(int \fIerrcode\fP, const regex_t *\fIpreg\fP,"
.B "     char *\fIerrbuf\fP, size_t \fIerrbuf_size\fP);"
.sp
.B void pcre2_regfree(regex_t *\fIpreg\fP);
.fi
.
.SH DESCRIPTION
.rs
.sp
This set of functions provides a POSIX-style API for the PCRE2 regular
expression 8-bit library. There are no POSIX-style wrappers for PCRE2's 16-bit
and 32-bit libraries. See the
.\" HREF
\fBpcre2api\fP
.\"
documentation for a description of PCRE2's native API, which contains much
additional functionality.
.P
\fBIMPORTANT NOTE\fP: The functions described here are NOT thread-safe, and
should not be used in multi-threaded applications. They are also limited to
processing subjects that are not bigger than 2GB. Use the native API instead.
.P
These functions are wrapper functions that ultimately call the PCRE2 native
API. Their prototypes are defined in the \fBpcre2posix.h\fP header file, and
they all have unique names starting with \fBpcre2_\fP. However, the
\fBpcre2posix.h\fP header also contains macro definitions that convert the
standard POSIX names such \fBregcomp()\fP into \fBpcre2_regcomp()\fP etc. This
means that a program can use the usual POSIX names without running the risk of
accidentally linking with POSIX functions from a different library.
.P
On Unix-like systems the PCRE2 POSIX library is called \fBlibpcre2-posix\fP, so
can be accessed by adding \fB-lpcre2-posix\fP to the command for linking an
application. Because the POSIX functions call the native ones, it is also
necessary to add \fB-lpcre2-8\fP.
.P
On Windows systems, if you are linking to a DLL version of the library, it is
recommended that \fBPCRE2POSIX_SHARED\fP is defined before including the
\fBpcre2posix.h\fP header, as it will allow for a more efficient way to
invoke the functions by adding the \fB__declspec(dllimport)\fP decorator.
.P
Although they were not defined as prototypes in \fBpcre2posix.h\fP, releases
10.33 to 10.36 of the library contained functions with the POSIX names
\fBregcomp()\fP etc. These simply passed their arguments to the PCRE2
functions. These functions were provided for backwards compatibility with
earlier versions of PCRE2, which had only POSIX names. However, this has proved
troublesome in situations where a program links with several libraries, some of
which use PCRE2's POSIX interface while others use the real POSIX functions.
For this reason, the POSIX names have been removed since release 10.37.
.P
Calling the header file \fBpcre2posix.h\fP avoids any conflict with other POSIX
libraries. It can, of course, be renamed or aliased as \fBregex.h\fP, which is
the "correct" name, if there is no clash. It provides two structure types,
\fIregex_t\fP for compiled internal forms, and \fIregmatch_t\fP for returning
captured substrings. It also defines some constants whose names start with
"REG_"; these are used for setting options and identifying error codes.
.
.
.SH "USING THE POSIX FUNCTIONS"
.rs
.sp
Note that these functions are just POSIX-style wrappers for PCRE2's native API.
They do not give POSIX regular expression behaviour, and they are not
thread-safe or even POSIX compatible.
.P
Those POSIX option bits that can reasonably be mapped to PCRE2 native options
have been implemented. In addition, the option REG_EXTENDED is defined with the
value zero. This has no effect, but since programs that are written to the
POSIX interface often use it, this makes it easier to slot in PCRE2 as a
replacement library. Other POSIX options are not even defined.
.P
There are also some options that are not defined by POSIX. These have been
added at the request of users who want to make use of certain PCRE2-specific
features via the POSIX calling interface or to add BSD or GNU functionality.
.P
When PCRE2 is called via these functions, it is only the API that is POSIX-like
in style. The syntax and semantics of the regular expressions themselves are
still those of Perl, subject to the setting of various PCRE2 options, as
described below. "POSIX-like in style" means that the API approximates to the
POSIX definition; it is not fully POSIX-compatible, and in multi-unit encoding
domains it is probably even less compatible.
.P
The descriptions below use the actual names of the functions, but, as described
above, the standard POSIX names (without the \fBpcre2_\fP prefix) may also be
used.
.
.
.SH "COMPILING A PATTERN"
.rs
.sp
The function \fBpcre2_regcomp()\fP is called to compile a pattern into an
internal form. By default, the pattern is a C string terminated by a binary
zero (but see REG_PEND below). The \fIpreg\fP argument is a pointer to a
\fBregex_t\fP structure that is used as a base for storing information about
the compiled regular expression. It is also used for input when REG_PEND is
set. The \fBregex_t\fP structure used by \fBpcre2_regcomp()\fP is defined in
\fBpcre2posix.h\fP and is not the same as the structure used by other libraries
that provide POSIX-style matching.
.P
The argument \fIcflags\fP is either zero, or contains one or more of the bits
defined by the following macros:
.sp
  REG_DOTALL
.sp
The PCRE2_DOTALL option is set when the regular expression is passed for
compilation to the native function. Note that REG_DOTALL is not part of the
POSIX standard.
.sp
  REG_ICASE
.sp
The PCRE2_CASELESS option is set when the regular expression is passed for
compilation to the native function.
.sp
  REG_NEWLINE
.sp
The PCRE2_MULTILINE option is set when the regular expression is passed for
compilation to the native function. Note that this does \fInot\fP mimic the
defined POSIX behaviour for REG_NEWLINE (see the following section).
.sp
  REG_NOSPEC
.sp
The PCRE2_LITERAL option is set when the regular expression is passed for
compilation to the native function. This disables all meta characters in the
pattern, causing it to be treated as a literal string. The only other options
that are allowed with REG_NOSPEC are REG_ICASE, REG_NOSUB, REG_PEND, and
REG_UTF. Note that REG_NOSPEC is not part of the POSIX standard.
.sp
  REG_NOSUB
.sp
When a pattern that is compiled with this flag is passed to
\fBpcre2_regexec()\fP for matching, the \fInmatch\fP and \fIpmatch\fP arguments
are ignored, and no captured strings are returned. Versions of the PCRE2 library
prior to 10.22 used to set the PCRE2_NO_AUTO_CAPTURE compile option, but this
no longer happens because it disables the use of backreferences.
.sp
  REG_PEND
.sp
If this option is set, the \fBreg_endp\fP field in the \fIpreg\fP structure
(which has the type const char *) must be set to point to the character beyond
the end of the pattern before calling \fBpcre2_regcomp()\fP. The pattern itself
may now contain binary zeros, which are treated as data characters. Without
REG_PEND, a binary zero terminates the pattern and the \fBre_endp\fP field is
ignored. This is a GNU extension to the POSIX standard and should be used with
caution in software intended to be portable to other systems.
.sp
  REG_UCP
.sp
The PCRE2_UCP option is set when the regular expression is passed for
compilation to the native function. This causes PCRE2 to use Unicode properties
when matching \ed, \ew, etc., instead of just recognizing ASCII values. Note
that REG_UCP is not part of the POSIX standard.
.sp
  REG_UNGREEDY
.sp
The PCRE2_UNGREEDY option is set when the regular expression is passed for
compilation to the native function. Note that REG_UNGREEDY is not part of the
POSIX standard.
.sp
  REG_UTF
.sp
The PCRE2_UTF option is set when the regular expression is passed for
compilation to the native function. This causes the pattern itself and all data
strings used for matching it to be treated as UTF-8 strings. Note that REG_UTF
is not part of the POSIX standard.
.P
In the absence of these flags, no options are passed to the native function.
This means that the regex is compiled with PCRE2 default semantics. In
particular, the way it handles newline characters in the subject string is the
Perl way, not the POSIX way. Note that setting PCRE2_MULTILINE has only
\fIsome\fP of the effects specified for REG_NEWLINE. It does not affect the way
newlines are matched by the dot metacharacter (they are not) or by a negative
class such as [^a] (they are).
.P
The yield of \fBpcre2_regcomp()\fP is zero on success, and non-zero otherwise.
The \fIpreg\fP structure is filled in on success, and one other member of the
structure (as well as \fIre_endp\fP) is public: \fIre_nsub\fP contains the
number of capturing subpatterns in the regular expression. Various error codes
are defined in the header file.
.P
NOTE: If the yield of \fBpcre2_regcomp()\fP is non-zero, you must not attempt
to use the contents of the \fIpreg\fP structure. If, for example, you pass it
to \fBpcre2_regexec()\fP, the result is undefined and your program is likely to
crash.
.
.
.SH "MATCHING NEWLINE CHARACTERS"
.rs
.sp
This area is not simple, because POSIX and Perl take different views of things.
It is not possible to get PCRE2 to obey POSIX semantics, but then PCRE2 was
never intended to be a POSIX engine. The following table lists the different
possibilities for matching newline characters in Perl and PCRE2:
.sp
                          Default   Change with
.sp
  . matches newline          no     PCRE2_DOTALL
  newline matches [^a]       yes    not changeable
  $ matches \en at end        yes    PCRE2_DOLLAR_ENDONLY
  $ matches \en in middle     no     PCRE2_MULTILINE
  ^ matches \en in middle     no     PCRE2_MULTILINE
.sp
This is the equivalent table for a POSIX-compatible pattern matcher:
.sp
                          Default   Change with
.sp
  . matches newline          yes    REG_NEWLINE
  newline matches [^a]       yes    REG_NEWLINE
  $ matches \en at end        no     REG_NEWLINE

NEWS
News about PCRE2 releases
-------------------------

Version 10.49 28-September-2026
-------------------------------

This is a security-only release, to address GHSA-r9hj-j2rw-4q3m.

Compared to 10.48, this release has only a minimal code change to prevent an
out-of-bounds write with arbitrary data. An attacker-controlled regex pattern
is required, and applications are only affected if using the
pcre2_jit_stack_create() and pcre2_jit_stack_assign() APIs to provide a growable
JIT stack, and then matching against a pattern with unusually high JIT stack
usage, such as a large number of capturing groups.

The implications of an out-of-bounds write could include arbitrary code
execution.

The issue is not a regression and affects releases 10.48 and earlier.


Version 10.48 31-August-2026
----------------------------

This is a regular release, incorporating security fixes along with small
improvements and fixes to library behaviour.

Only changes to behaviour, changes to the API, and other significant changes
are described here. Please see the ChangeLog and Git log for further details.

As well as the tarball and Git tag for this release, there are detailed
instructions for backporting security and correctness fixes, for the last
five years of releases.

* (Git change) Renamed the default development branch from master to main.

* (Maintenance change) Added a five-year support lifecycle policy and
publication of backport patches for security and high-severity fixes in older
releases.

* (Security fix for very specific API usage, GHSA-2p8c-ff85-vh9x) If
pcre2_jit_compile() is called with options for some match modes, and then
pcre2_match() is used to perform a match for a different match mode, an
out-of-bounds read can occur if the match is attempted against invalid UTF input.

* (Security fix for pattern conversion, GHSA-q8g2-wprr-34m9) If pcre2_convert()
is called on untrusted input on platforms with 32-bit size_t, an out-of-bounds
heap write can occur.

* (Security fix, GHSA-3r4p-g7gg-ppmf) Fixed an out-of-bounds write in DFA
matching when using a heap limit; also fixed possible integer overflows which
could cause under-allocation of the workspace.

* (Security fix, GHSA-fmgr-6ggq-9859) Added bounds checks for several integer
overflows while compiling patterns on 32-bit CPUs, which could cause
under-allocation followed by out-of-bounds writes.

* (Security fix, GHSA-9qww-pwc4-77qq) Applied lower buffer bound to prevent
two out-of-bounds reads while scanning backwards through invalid UTF data with
PCRE2_MATCH_INVALID_UTF.

* (Matching correctness) Fixed several matching issues:
  - A JIT-specific matching bug affecting prefix scanning on patterns with
    repeats (#875).
  - A JIT-specific matching bug in variable-length lookbehinds (#912).
  - Miscompiled Unicode character classes combining characters at or below
    U+00FF with characters at U+0100 and U+8000 or above (#841).
  - Incorrect JIT character advancement with PCRE2_MATCH_INVALID_UTF in UTF-8
    and UTF-16 modes, which could skip adjacent characters (#945).

* (Behaviour change) Updated Unicode support to Unicode 17.0.

* (Small behaviour changes) Many small fixes, including pcre2_substitute()
improvements, optimisation of possessive backreference matching, and
pcre2_compile() fixes.

* (Small build changes) Many small adjustments to the CMake and Zig builds.

* (Security fix for very specific API usage, #937) Fixed a leak and later
invalid free when calling the fast-path pcre2_jit_match() function with a match
data object previously used with pcre2_match() and
PCRE2_COPY_MATCHED_SUBJECT.

* (Low-severity security fix, GHSA-q7rw-r7qq-2hx6) Fixed exposure of two
uninitialised bytes from malloc() via pcre2_serialize_encode().


Version 10.47 21-October-2025
-----------------------------

This is a regular semi-annual release, incorporating a few new features and
several maintenance and build improvements.

Only changes to behaviour, changes to the API, and other significant changes
are described here. Please see the ChangeLog and Git log for further details.

* (Powerful new feature) Pattern recursion of the form
"(?1(GROUP_NAME_OR_NUM,...))" acts as a subroutine call which additionally
returns the listed capturing groups to the calling context.

* (Significant bugfix) Fixed a crash in pcre2_callout_enumerate() which is
easily reachable on any pattern that contains a Unicode character class. If your
application uses this function, please read the details for this change and
evaluate its severity for your application.

* (Build change) There are now linker scripts to enable symbol versioning for
the PCRE2 dynamic libraries. Downstream Linux distributions may make use of
this, or disable it with the new Autoconf `--disable-symvers` and CMake
`-DPCRE2_SYMVERS` options. Linux, Solaris, and FreeBSD (GNU ld, LLVM lld, and
Solaris ld) are tested and supported.

* (New API function) Added pcre2_next_match(). This function makes it both
simpler and safer for clients to iterate over all matches in a subject. The

src/pcre2posix.c
/*************************************************
*      Perl-Compatible Regular Expressions       *
*************************************************/

/* PCRE is a library of functions to support regular expressions whose syntax
and semantics are as close as possible to those of the Perl 5 language.

                       Written by Philip Hazel
     Original API code Copyright (c) 1997-2012 University of Cambridge
          New API code Copyright (c) 2016-2024 University of Cambridge

-----------------------------------------------------------------------------
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

    * Neither the name of the University of Cambridge nor the names of its
      contributors may be used to endorse or promote products derived from
      this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
-----------------------------------------------------------------------------
*/


/* This module is a wrapper that provides a POSIX API to the underlying PCRE2
functions. The functions are called pcre2_regcomp(), pcre2_regexec(), etc.
pcre2posix.h defines the POSIX names as macros for the corresponding pcre2_xxx
functions, so any program that includes it and uses the POSIX names will call
the PCRE2 implementations instead. */


/* This module doesn't use pcre2_internal.h, because the pcre2posix dynamic
library has an "internal" view of some macros, but is an "external" client of
the pcre2-8 dynamic library. This is unusual, and justifies a (rare) direct
inclusion of config.h. */

#if defined HAVE_CONFIG_H && !defined PCRE2_CONFIG_H_IDEMPOTENT_GUARD
#define PCRE2_CONFIG_H_IDEMPOTENT_GUARD
#include "config.h"
#endif



/* Ensure that the PCRE2POSIX_EXP_xxx macros are set appropriately for
compiling these functions. This must come before including pcre2posix.h, where
they are set for an application (using these functions) if they have not
previously been set. */

#if defined __cplusplus
#error This project uses C99. C++ is not supported.
#endif

#ifndef PCRE2POSIX_EXP_DECL
#if !defined(PCRE2POSIX_SHARED)
#define PCRE2POSIX_EXP_DECL extern
#elif defined(_WIN32)
#define PCRE2POSIX_EXP_DECL extern __declspec(dllexport)
#else
#define PCRE2POSIX_EXP_DECL extern PCRE2_EXPORT
#endif
#endif

#ifndef PCRE2POSIX_EXP_DEFN
#if !defined(PCRE2POSIX_SHARED)
#define PCRE2POSIX_EXP_DEFN extern
#elif defined(_WIN32)
#define PCRE2POSIX_EXP_DEFN extern __declspec(dllexport)
#else
#define PCRE2POSIX_EXP_DEFN extern PCRE2_EXPORT
#endif
#endif

/* Older versions of MSVC lack snprintf(). This define allows for
warning/error-free compilation and testing with MSVC compilers back to at least
MSVC 10/2010. Except for VC6 (which is missing some fundamentals and fails). */

#if defined(_MSC_VER) && (_MSC_VER < 1900)
#define snprintf _snprintf
#endif


/* Compile-time error numbers start at this value. It should probably never be
changed. This #define is a copy of the one in pcre2_internal.h. */

#define COMPILE_ERROR_BASE 100


/* Standard C headers */

#include <ctype.h>
#include <limits.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* PCRE2 headers */

#include "pcre2.h"
#include "pcre2posix.h"
#include "pcre2_util.h"

/* For pcre2_tables.c */
#define PCRE2_PCRE2POSIX
/* Avoid collisions with pcre2test's tables when linking statically. */
#define PRIV(name) _pcre2posix_##name

#include "pcre2_tables.c"

/* Table to translate PCRE2 compile time error codes into POSIX error codes.
Only a few PCRE2 errors with a value greater than 23 turn into special POSIX
codes: most go to REG_BADPAT. The second table lists, in pairs, those that
don't, even though some of them cannot currently be provoked from within the
POSIX wrapper. */

static const int eint1[] = {
  // clang-format off
  0,           /* No error */
  REG_EESCAPE, /* \ at end of pattern */
  REG_EESCAPE, /* \c at end of pattern */
  REG_EESCAPE, /* unrecognized character follows \ */
  REG_BADBR,   /* numbers out of order in {} quantifier */
  /* 5 */
  REG_BADBR,   /* number too big in {} quantifier */
  REG_EBRACK,  /* missing terminating ] for character class */
  REG_ECTYPE,  /* invalid escape sequence in character class */
  REG_ERANGE,  /* range out of order in character class */
  REG_BADRPT,  /* nothing to repeat */
  /* 10 */
  REG_ASSERT,  /* internal error: unexpected repeat */
  REG_BADPAT,  /* unrecognized character after (? or (?- */
  REG_BADPAT,  /* POSIX named classes are supported only within a class */
  REG_BADPAT,  /* POSIX collating elements are not supported */
  REG_EPAREN,  /* missing ) */
  /* 15 */
  REG_ESUBREG, /* reference to non-existent subpattern */
  REG_INVARG,  /* pattern passed as NULL */
  REG_INVARG,  /* unknown compile-time option bit(s) */
  REG_EPAREN,  /* missing ) after (?# comment */
  REG_ESIZE,   /* parentheses nested too deeply */
  /* 20 */
  REG_ESIZE,   /* regular expression too large */
  REG_ESPACE,  /* failed to get memory */
  REG_EPAREN,  /* unmatched closing parenthesis */
  REG_ASSERT,  /* internal error: code overflow */
  // clang-format on
};

static const int eint2[] = {
  // clang-format off
   30, REG_ECTYPE,  /* unknown POSIX class name */
   32, REG_INVARG,  /* this version of PCRE2 does not have Unicode support */
   37, REG_EESCAPE, /* PCRE2 does not support \L, \l, \N{name}, \U, or \u */
   56, REG_INVARG,  /* internal error: unknown newline setting */
   92, REG_INVARG,  /* invalid option bits with PCRE2_LITERAL */
   98, REG_EESCAPE, /* missing digit after \0 in NO_BS0 mode */
   99, REG_EESCAPE, /* \K in lookaround */
  102, REG_EESCAPE, /* \ddd octal > \377 in PYTHON_OCTAL mode */
  // clang-format on
};

/* Table of texts corresponding to POSIX error codes */

static const char *const pstring[] = {
  // clang-format off
  "",                                /* Dummy for value 0 */
  "internal error",                  /* REG_ASSERT */
  "invalid repeat counts in {}",     /* BADBR      */
  "pattern error",                   /* BADPAT     */
  "? * + invalid",                   /* BADRPT     */
  "unbalanced {}",                   /* EBRACE     */
  "unbalanced []",                   /* EBRACK     */
  "collation error - not relevant",  /* ECOLLATE   */
  "bad class",                       /* ECTYPE     */
  "bad escape sequence",             /* EESCAPE    */
  "empty expression",                /* EMPTY      */
  "unbalanced ()",                   /* EPAREN     */
  "bad range inside []",             /* ERANGE     */
  "expression too big",              /* ESIZE      */
  "failed to get memory",            /* ESPACE     */
  "bad back reference",              /* ESUBREG    */
  "bad argument",                    /* INVARG     */
  "match failed",                    /* NOMATCH    */
  // clang-format on
};

/*************************************************
*          Translate error code to string        *
*************************************************/

/* Ensure that re_erroffset can be passed to snprintf() as %llu. */
STATIC_ASSERT(SIZE_MAX <= ULLONG_MAX, size_t_fits_in_unsigned_long_long);

PCRE2POSIX_EXP_DEFN size_t PCRE2_CALL_CONVENTION
pcre2_regerror(int errcode, const regex_t *preg, char *errbuf, size_t errbuf_size)
{
  const char *message;
  /* Ensure that offset_buf has enough space for a size_t decimal. This estimate
  of N characters in an N-bit build is a gross overestimate, but safe. */
  char offset_buf[sizeof(" at offset ") + sizeof(size_t) * CHAR_BIT];
  int snprintf_rc, have_offset = 0;
  PCRE2_SIZE i;

  message = (errcode <= 0 || errcode >= (int)(sizeof(pstring) / sizeof(char *)))
                ? "unknown error code"
                : pstring[errcode];

  if (preg != NULL &&
      preg->re_erroffset != SIZE_MAX
      // LCOV_EXCL_START - snprintf failures here are essentially unreachable
      && (snprintf_rc = snprintf(offset_buf, sizeof(offset_buf), " at offset %llu",
                                 (unsigned long long)preg->re_erroffset)) > 0 &&
      snprintf_rc < (int)sizeof(offset_buf)
      // LCOV_EXCL_STOP
  )
  {
    have_offset = 1;
    offset_buf[sizeof(offset_buf) - 1] = 0; // Paranoia for very old snprintf
  }

  for (i = 0; *message != 0; i++, message++)
    if (i + 1 < errbuf_size)
      errbuf[i] = *message;

  if (have_offset)
  {
    for (message = offset_buf; *message != 0; i++, message++)
      if (i + 1 < errbuf_size)
        errbuf[i] = *message;
  }

#if defined EBCDIC && 'a' != 0x81
  /* If compiling for EBCDIC, but the compiler's string literals are not EBCDIC,
  then we are in the "force EBCDIC 1047" mode. I have chosen to add a few lines
  here to translate the error strings on the fly, rather than require the string
  literals above to be written out arduously using the "STR_XYZ" macros. */
  for (PCRE2_SIZE j = 0; j < i && j + 1 < errbuf_size; ++j)
    errbuf[j] = PRIV(ascii_to_ebcdic_1047)[(uint8_t)errbuf[j]];
#endif

  /* Terminate message, even if truncated. */

  if (errbuf_size > 0)
    errbuf[(i < errbuf_size) ? i : errbuf_size - 1] = 0;
  i++;

  return i;
}



/*************************************************
*           Free store held by a regex           *
*************************************************/

PCRE2POSIX_EXP_DEFN void PCRE2_CALL_CONVENTION
pcre2_regfree(regex_t *preg)
{
  if (preg == NULL)
    return;
  pcre2_match_data_free(preg->re_match_data);
  preg->re_match_data = NULL;
  pcre2_code_free(preg->re_pcre2_code);
  preg->re_pcre2_code = NULL;
}



/*************************************************
*            Compile a regular expression        *
*************************************************/

/*
Arguments:
  preg        points to a structure for recording the compiled expression
  pattern     the pattern to compile
  cflags      compilation flags

Returns:      0 on success
              various non-zero codes on failure
*/

PCRE2POSIX_EXP_DEFN int PCRE2_CALL_CONVENTION
pcre2_regcomp(regex_t *preg, const char *pattern, int cflags)
{
  PCRE2_SIZE erroffset;
  PCRE2_SIZE patlen;
  int errorcode;
  pcre2_code *re_pcre2_code;
  pcre2_match_data *re_match_data;
  uint32_t options = 0;
  uint32_t re_nsub = 0;

  if (preg == NULL)
    return REG_INVARG;

  preg->re_match_data = NULL;
  preg->re_pcre2_code = NULL;
  preg->re_nsub = 0;
  preg->re_erroffset = SIZE_MAX;
  preg->re_cflags = 0;

  if (pattern == NULL)
    return REG_INVARG;

  if ((cflags & REG_PEND) != 0)
  {
    if (preg->re_endp == NULL || (uintptr_t)preg->re_endp < (uintptr_t)pattern)
      return REG_INVARG;
    patlen = (PCRE2_SIZE)(preg->re_endp - pattern);
  }
  else
  {
    preg->re_endp = NULL;
    patlen = PCRE2_ZERO_TERMINATED;
  }

  if ((cflags & ~(REG_ICASE | REG_NEWLINE | REG_DOTALL | REG_NOSUB | REG_UTF | REG_UCP |
                  REG_UNGREEDY | REG_PEND | REG_NOSPEC)) != 0)
    return REG_INVARG;

  if ((cflags & REG_ICASE) != 0)
    options |= PCRE2_CASELESS;
  if ((cflags & REG_NEWLINE) != 0)
    options |= PCRE2_MULTILINE;
  if ((cflags & REG_DOTALL) != 0)
    options |= PCRE2_DOTALL;
  if ((cflags & REG_NOSPEC) != 0)
    options |= PCRE2_LITERAL;
  if ((cflags & REG_UTF) != 0)
    options |= PCRE2_UTF;
  if ((cflags & REG_UCP) != 0)
    options |= PCRE2_UCP;
  if ((cflags & REG_UNGREEDY) != 0)
    options |= PCRE2_UNGREEDY;

  re_pcre2_code = pcre2_compile((PCRE2_SPTR)pattern, patlen, options, &errorcode, &erroffset, NULL);

  if (re_pcre2_code == NULL)
  {
    unsigned int i;

    preg->re_erroffset = erroffset;

    /* A negative value is a UTF error; otherwise all error codes are greater
    than COMPILE_ERROR_BASE, but check, just in case. */

    if (errorcode < COMPILE_ERROR_BASE)
      return REG_BADPAT;
    errorcode -= COMPILE_ERROR_BASE;

    if (errorcode < (int)(sizeof(eint1) / sizeof(const int)))
      return eint1[errorcode];
    for (i = 0; i < sizeof(eint2) / sizeof(const int); i += 2)
      if (errorcode == eint2[i])
        return eint2[i + 1];
    return REG_BADPAT;
  }

  (void)pcre2_pattern_info(re_pcre2_code, PCRE2_INFO_CAPTURECOUNT, &re_nsub);
  re_match_data = pcre2_match_data_create(re_nsub + 1, NULL);

  if (re_match_data == NULL)
  {
    /* There is no facility for passing a custom allocator to the POSIX API, so
    our test code cannot force a malloc failure here. If there were an API to
    customize the default (global) PCRE2 allocator, we could test it. Since the
    code is nonetheless reachable, I prefer not to exclude it from coverage
    reporting. */
    pcre2_code_free(re_pcre2_code);
    return REG_ESPACE;
  }

  preg->re_pcre2_code = re_pcre2_code;
  preg->re_match_data = re_match_data;
  preg->re_nsub = (size_t)re_nsub;
  preg->re_cflags = cflags;

  return 0;
}



/*************************************************
*              Match a regular expression        *
*************************************************/

/* A match_data block was obtained when the pattern was compiled, to save having
to allocate and free it for each match. If REG_NOSUB was specified at compile
time, the nmatch and pmatch arguments are ignored except as input for
REG_STARTEND, and the only result is yes/no/error. */

PCRE2POSIX_EXP_DEFN int PCRE2_CALL_CONVENTION
pcre2_regexec(const regex_t *preg, const char *string, size_t nmatch, regmatch_t pmatch[],
              int eflags)
{
  int rc;
  PCRE2_SIZE so, eo;
  uint32_t options = 0;
  pcre2_match_data *md;

  if (preg == NULL || preg->re_pcre2_code == NULL || preg->re_match_data == NULL || string == NULL)
    return REG_INVARG;

  if ((eflags & ~(REG_NOTBOL | REG_NOTEOL | REG_NOTEMPTY | REG_STARTEND)) != 0)
    return REG_INVARG;

  md = (pcre2_match_data *)preg->re_match_data;

  if ((eflags & REG_NOTBOL) != 0)
    options |= PCRE2_NOTBOL;
  if ((eflags & REG_NOTEOL) != 0)
    options |= PCRE2_NOTEOL;
  if ((eflags & REG_NOTEMPTY) != 0)
    options |= PCRE2_NOTEMPTY;

  /* When REG_NOSUB was specified, or if no vector has been passed in which to
  put captured strings, ensure that nmatch is zero. This will stop any attempt to
  write to pmatch. */

  if ((preg->re_cflags & REG_NOSUB) != 0 || pmatch == NULL)
    nmatch = 0;

  /* REG_STARTEND is a BSD extension, to allow for non-NUL-terminated strings.
  The man page from OS X says "REG_STARTEND affects only the location of the
  string, not how it is matched". That is why the "so" value is used to bump the
  start location rather than being passed as a PCRE2 "starting offset". */

  if ((eflags & REG_STARTEND) != 0)
  {
    if (pmatch == NULL)
      return REG_INVARG;
    if (pmatch[0].rm_so < 0 || pmatch[0].rm_eo < 0 || pmatch[0].rm_so > pmatch[0].rm_eo)
      return REG_INVARG;
    so = (PCRE2_SIZE)pmatch[0].rm_so;
    eo = (PCRE2_SIZE)pmatch[0].rm_eo;
  }
  else
  {
    so = 0;
    eo = strlen(string);
  }

  rc = pcre2_match((const pcre2_code *)preg->re_pcre2_code, (PCRE2_SPTR)string + so, (eo - so), 0,
                   options, md, NULL);

  /* Successful match */

  if (rc >= 0)
  {
    size_t i;
    size_t count = rc == 0 ? pcre2_get_ovector_count(md) : (size_t)rc;
    PCRE2_SIZE *ovector = pcre2_get_ovector_pointer(md);
    if (count > nmatch)
      count = nmatch;
    for (i = 0; i < count; i++)
    {
      if (ovector[i * 2] == PCRE2_UNSET || ovector[i * 2] + so > (PCRE2_SIZE)INT_MAX ||
          ovector[i * 2 + 1] == PCRE2_UNSET || ovector[i * 2 + 1] + so > (PCRE2_SIZE)INT_MAX)
      {
        pmatch[i].rm_so = -1;
        pmatch[i].rm_eo = -1;
      }
      else
      {
        pmatch[i].rm_so = (int)(ovector[i * 2] + so);
        pmatch[i].rm_eo = (int)(ovector[i * 2 + 1] + so);
      }
    }

    for (; i < nmatch; i++)
      pmatch[i].rm_so = pmatch[i].rm_eo = -1;
    return 0;
  }

  /* Unsuccessful match */

  if (rc <= PCRE2_ERROR_UTF8_ERR1 && rc >= PCRE2_ERROR_UTF8_ERR21)
    return REG_INVARG;

  /* Most of these are events that won't occur during testing, so exclude them
  from coverage. */

  switch (rc)
  {
  case PCRE2_ERROR_BAD_BACKSLASH_K:
    return REG_BADPAT;
  case PCRE2_ERROR_DEPTHLIMIT:
    return REG_ESPACE;
  case PCRE2_ERROR_HEAPLIMIT:
    return REG_ESPACE;
  case PCRE2_ERROR_MATCHLIMIT:
    return REG_ESPACE;
  case PCRE2_ERROR_NOMATCH:
    return REG_NOMATCH;
  case PCRE2_ERROR_RECURSELOOP:
    return REG_BADPAT;

    /* LCOV_EXCL_START */
  case PCRE2_ERROR_BADMODE:
    return REG_INVARG;
  case PCRE2_ERROR_BADMAGIC:
    return REG_INVARG;
  case PCRE2_ERROR_BADOFFSET:
    return REG_INVARG;
  case PCRE2_ERROR_BADOFFSETLIMIT:
    return REG_INVARG;
  case PCRE2_ERROR_BADOPTION:
    return REG_INVARG;
  case PCRE2_ERROR_BADUTFOFFSET:
    return REG_INVARG;
  case PCRE2_ERROR_JIT_STACKLIMIT:
    return REG_ESPACE;
  case PCRE2_ERROR_NOMEMORY:
    return REG_ESPACE;
  case PCRE2_ERROR_NULL:
    return REG_INVARG;
  default:
    return REG_ASSERT;
    /* LCOV_EXCL_STOP */
  }
}

/* End of pcre2posix.c */
